### ISSUE TITLE
TBC Downranking Behaviour

### ISSUE THREAD

**Issue by kaytotes** [Project-Epoch/bugtracker#2600]
**Spell ID's:** 5188, 9889

> +Healing isnt working for any healing touch ranks lower than 7.

> Cannot cast this in tree of life form. Bug or working as intended?

**Comment by zana244** [comment]
first issue is confirmed on internal... I'm also not sure if the full 100% coefficient is being applied to the ranks >7...
I only looked at coefficients on TBC wowhead, unsure if they are supposed to be different.
https://www.wowhead.com/tbc/spell=9888/healing-touch#comments
second one is a non-issue I think, the wording on the spell is clear...

**Comment by Shadownflame** [comment]
Probably not an issue unique to healing touch, epoch hasn't implemented tbc downranking system.

**Comment by kaytotes** [comment]
Next deploy

**Comment by zana244** [comment]
Coefficients still seem mangled?
Healing touch for druid has 0 coefficient even at rank 8 now... only rank 9 and rank 10.
Rejuv at rank 8 also has 0 coefficient.

Lightning bolt behaves in a similar way... something is iffy (unsure if related to this though).

**Comment by kaytotes** [comment]
I officially clearly don't know where this is handled core side.

**Comment by Shadownflame** [comment]
![Image](https://github.com/user-attachments/assets/c456a413-16ce-429f-ac08-993a6f8ca911)

**Comment by ulmetrs** [comment]
I will look at this soonish, maybe kids nap time and if not tonight

**Comment by T1ti** [comment]
https://github.com/Project-Epoch/TrinityCore/blob/epoch-core/src/server/game/Entities/Unit/Unit.cpp#L2483

This ? I see you added a ts hook but it happens after so it would only work if you redo the calculation

**Comment by kaytotes** [comment]
We are using that hook. Issue must be elsewhere before it.

**Comment by T1ti** [comment]
Verify in that function that `result` after the hook is correct, I don't know what you're doing with it

other than that `CalculateDefaultCoefficient`() is usualled called right before `CalculateSpellpowerCoefficientLevelPenalty`() and both are multiplied together

**Comment by kaytotes** [comment]
Can replicate it going to 0 but formula matches cMangos and old TC reference before they changed it so gonna slap in some prints and see what I can work out.

**Comment by kaytotes** [comment]
tl;dr int float conversion shenanigans. I've validated with healing touch rank 10 to 1 descending.

```
Spell: 9889 - Level: 56 - Max Level: 61 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 1.000000
Stock Penalty: 1.000000 - Final Penalty: 1.000000

Spell: 9888 - Level: 50 - Max Level: 55 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 1.000000
Stock Penalty: 0.850000 - Final Penalty: 1.000000

Spell: 9758 - Level: 44 - Max Level: 49 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 0.916667
Stock Penalty: 0.550000 - Final Penalty: 0.916667

Spell: 8903 - Level: 38 - Max Level: 43 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 0.816667
Stock Penalty: 0.250000 - Final Penalty: 0.816667

Spell: 6778 - Level: 32 - Max Level: 37 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 0.716667
Stock Penalty: 0.000000 - Final Penalty: 0.716667

Spell: 5189 - Level: 26 - Max Level: 31 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 0.616667
Stock Penalty: 0.000000 - Final Penalty: 0.616667

Spell: 5188 - Level: 20 - Max Level: 25 - Player Level: 60
Level Penalty: 0.000000 - Level Factor: 0.516667
Stock Penalty: 0.000000 - Final Penalty: 0.516667

Spell: 5187 - Level: 14 - Max Level: 19 - Player Level: 60
Level Penalty: 22.500000 - Level Factor: 0.416667
Stock Penalty: 0.000000 - Final Penalty: 0.322917

Spell: 5186 - Level: 8 - Max Level: 13 - Player Level: 60
Level Penalty: 45.000000 - Level Factor: 0.316667
Stock Penalty: 0.000000 - Final Penalty: 0.174167

Spell: 5185 - Level: 1 - Max Level: 5 - Player Level: 60
Level Penalty: 71.250000 - Level Factor: 0.183333
Stock Penalty: 0.000000 - Final Penalty: 0.052708
```

**Comment by kaytotes** [comment]
Deployed
